feat(runner): add metadata parameter to Runner.run_async()

donggyun112 opened the pull request:
---
  Add metadata parameter to Runner.run_async()

  Add support for passing per-request metadata through the agent execution pipeline. This enables use cases like:
  - Passing user_id, trace_id, or session context to callbacks
  - Passing context keys for memory injection in before_model_callback

  Changes:
  - Add metadata field to LlmRequest model
  - Add metadata field to InvocationContext model
  - Add metadata parameter to Runner.run_async() and related methods
  - Propagate metadata from InvocationContext to LlmRequest in base_llm_flow
  - Add unit tests for metadata functionality

  Closes #2978

  ---
  Link to Issue or Description of Change

  1. Link to an existing issue:

  - Closes: #2978

  2. Description:

  Problem:
  Currently, there is no official way to pass per-request metadata (such as user_id, trace_id, or memory context keys) from Runner.run_async() to callbacks like before_model_callback.
  
  Solution:
  Add a metadata parameter to Runner.run_async() that flows through InvocationContext to LlmRequest. This provides a clean, official API for passing request-specific context to callbacks without workarounds.

  Testing Plan

  Unit Tests:

  - I have added or updated unit tests for my change.
  - All unit tests pass locally.

  ======================= 35 passed, 22 warnings in 2.79s ========================

  Test cases added:
  - test_new_invocation_context_with_metadata
  - test_new_invocation_context_without_metadata
  - test_run_async_passes_metadata_to_invocation_context
  - test_metadata_field_in_invocation_context
  - test_metadata_field_in_llm_request
  - test_llm_request_without_metadata

  Manual End-to-End Tests:
```python
  from google.adk.runners import Runner

  # Pass metadata when calling run_async
  async for event in runner.run_async(
      user_id="user123",
      session_id="session456",
      new_message=content,
      metadata={"experiment_id": "exp-001", "variant": "B"}
  ):
      # metadata is accessible in before_model_callback via llm_request.metadata
      pass
```

  Checklist

  - [x] I have read the CONTRIBUTING.md document.
  - [x] I have performed a self-review of my own code.
  - [x] I have commented my code, particularly in hard-to-understand areas.
  - [x] I have added tests that prove my fix is effective or that my feature works.
  - [x] New and existing unit tests pass locally with my changes.
  - [x] I have manually tested my changes end-to-end.
  - [x] Any dependent changes have been merged and published in downstream modules.

  Additional context

  This PR addresses the need discussed in issue #2978 for a way to pass per-request context to callbacks.

donggyun112 committed:
feat(runner): add metadata parameter to Runner.run_async()

Add support for passing per-request metadata through the agent execution
pipeline. This enables use cases like:
- Passing user_id, trace_id, or session context to callbacks
- Enabling memory injection in before_model_callback
- Supporting request-specific context without using ContextVar workarounds

Changes:
- Add `metadata` field to LlmRequest model
- Add `metadata` field to InvocationContext model
- Add `metadata` parameter to Runner.run_async() and related methods
- Propagate metadata from InvocationContext to LlmRequest in base_llm_flow
- Add unit tests for metadata functionality

Closes #2978

donggyun112 committed:
docs: clarify shallow copy behavior in docstring and add isolation test

- Add note in docstring about shallow copy behavior for nested objects
- Add test_metadata_shallow_copy_isolation to verify:
  - Top-level changes are isolated from original dict
  - Nested object modifications affect original (shallow copy)

donggyun112 committed:
fix: preserve empty dict metadata instead of converting to None

- Change `if metadata` to `if metadata is not None` for truthiness check
- Empty dict {} was incorrectly converted to None due to falsy check
- Add test_empty_metadata_dict_not_converted_to_none to prevent regression

ryanaiagent commented:
Hi @wyf7107 , can you please review this.

donggyun112 committed:
feat(runner): add metadata parameter to run(), run_live(), run_debug()

Add metadata support to all run methods for consistency:
- run(): sync wrapper, passes metadata to run_async()
- run_live(): live mode, passes metadata through invocation context
- run_debug(): debug helper, passes metadata to run_async()

Also update InvocationContext docstring to reflect all supported entry points.

donggyun112 commented:
Updated to add metadata support to all Runner entry points for API consistency:

- `run()`: sync wrapper now passes metadata
- `run_live()`: passes metadata through invocation context
- `run_debug()`: debug helper passes metadata

Added 3 new tests to cover these entry points.

ryanaiagent commented:
Hi @donggyun112 , Can you please fix the failing unit tests and formatting errors

donggyun112 committed:
fix: resolve isort duplicate import and add missing live_request_queue argument in test

- Remove duplicate LiveRequestQueue import caused by merge conflict
- Fix import ordering for tests.unittests module
- Add required live_request_queue argument to test_new_invocation_context_for_live_with_metadata test

donggyun112 commented:
Found a bug in upstream: BigQuery plugin's `_atexit_cleanup` raises `ReferenceError` when weakref proxy object is already garbage collected. Added a fix to handle this exception.

donggyun112 committed:
fix: move LiveRequestQueue to top-level import to avoid CI failures

harshakokel commented:
Voting up on this. 

This feature is really essential for any custom metadata to be passed to agents.

donggyun112 committed:
test(runner): slim metadata tests and align upstream conventions

- Remove 5 redundant tests (Pydantic field/default validation, private
  helper unit tests already covered end-to-end)
- Migrate gemini-2.0-flash to gemini-2.5-flash (4 occurrences) per
  upstream ed8b31ce migration
- Switch testing_utils import to relative form per codebase convention
- Hoist asyncio import to module top-level
- Drop unused LlmRequest import

donggyun112 committed:
Merge branch 'main' into feat/add-metadata-parameter

donggyun112 committed:
Merge branch 'main' into feat/add-metadata-parameter

donggyun112 committed:
Merge branch 'main' into feat/add-metadata-parameter

donggyun112 committed:
Merge upstream/main into feat/add-metadata-parameter

Resolve runners.py conflicts caused by upstream v2.0.0 GA transition:
- run_async signature: keep both metadata (ours) and yield_user_message (upstream)
- run_async docstring: keep both descriptions
- aclosing/_run_with_trace: adopt upstream's lowercase aclosing while keeping
  metadata argument forwarding
- run_debug: adopt upstream's aclosing pattern while preserving metadata=metadata

Also wire metadata through new v2.0.0 _run_node_async helper so the parameter
reaches InvocationContext on the BaseNode code path (run_async -> _run_node_async
-> _new_invocation_context).

donggyun112 committed:
Merge branch 'main' into feat/add-metadata-parameter

donggyun112 committed:
Merge branch 'main' into feat/add-metadata-parameter

donggyun112 committed:
docs: wrap metadata docstrings to 80-char line limit

donggyun112 commented:
Hi @rohityan @RyanAIAgent, sorry to ping again. Just rebased this onto the latest main and sorted out the runner conflicts from the v2 changes. CLA's green now and everything passes on my end (mypy/lint/tests). The CI just needs a maintainer to kick off the workflow run. Would really appreciate a look when you get a chance — happy to fix anything. Thanks!

donggyun112 committed:
Merge upstream/main into feat/add-metadata-parameter

Resolves conflicts in runners.py (merge metadata kwarg with upstream's
invocation_id kwarg in _setup_context_for_new_invocation) and removes
duplicate test definitions in test_runners.py that upstream now provides
at module level.

donggyun112 committed:
Merge upstream/main into feat/add-metadata-parameter

Coexists our public metadata field with upstream's new _custom_metadata
PrivateAttr in InvocationContext.

donggyun112 committed:
Merge upstream/main into feat/add-metadata-parameter

Coexists our public metadata field with upstream's new _is_managed_agent
PrivateAttr in LlmRequest.

donggyun112 committed:
Merge branch 'main' into feat/add-metadata-parameter

donggyun112 committed:
Merge upstream/main into feat/add-metadata-parameter

Reconciles metadata/session kwargs on _run_node_async and adds
test_run_async_rejects_user_function_call alongside TestRunnerMetadata.

donggyun112 committed:
Merge remote-tracking branch 'upstream/main' into feat/add-metadata-parameter

donggyun112 committed:
chore: untrack local .codecanvas/ artifacts

These are lean-ctx cache files (74MB callgraph.json + entrypoints.json)
that were accidentally tracked. Remove from index and gitignore.

donggyun112 commented:
Hi @rohityan, checking in again. Rebased onto latest main a couple times since the last ping to clear new conflicts, and dropped a stray artifact file that had gotten committed. Tests still passing locally, CI checks green on the last runs. Would be great to get this on your queue when you have a moment. Thanks!

donggyun112 committed:
Merge remote-tracking branch 'upstream/main' into feat/add-metadata-parameter

# Conflicts:
#	src/google/adk/models/llm_request.py

donggyun112 committed:
Merge branch 'main' into feat/add-metadata-parameter

donggyun112 committed:
Merge branch 'main' into feat/add-metadata-parameter

donggyun112 committed:
Merge branch 'main' into feat/add-metadata-parameter

donggyun112 committed:
Merge branch 'main' into feat/add-metadata-parameter

donggyun112 committed:
Merge branch 'main' into feat/add-metadata-parameter

donggyun112 committed:
Merge branch 'main' into feat/add-metadata-parameter

donggyun112 committed:
Merge branch 'main' into feat/add-metadata-parameter

donggyun112 committed:
Merge branch 'main' into feat/add-metadata-parameter

donggyun112 committed:
Merge branch 'main' into feat/add-metadata-parameter

GWeale commented:
sorry, this sat for months after you were told it was under review, and thanks for the rebases. the use case is covered now, /run and /run_sse take custom_metadata and callbacks read it as callback_context.custom_metadata, so closing this.
